src/daft-recordbatch/src/ops/window.rs

```diff
@@ -69,10 +68,205 @@ impl RecordBatch {
+        // Create a sequence of row numbers (1-based)
+        let row_numbers: Vec<u64> = (1..=self.len() as u64).collect();
+        let row_number_series = UInt64Array::from((name.as_str(), row_numbers)).into_series();
+        let row_number_batch = Self::from_nonempty_columns(vec![row_number_series])?;
+
+        self.union(&row_number_batch)
+    }
+
+    pub fn window_rank(&self, name: String, order_by: &[ExprRef], dense: bool) -> DaftResult<Self> {
+        // Create rank numbers for pre-sorted partition
+        let mut rank_numbers = vec![0u64; self.len()];
+
+        if self.is_empty() {
+            let rank_series = UInt64Array::from((name.as_str(), rank_numbers)).into_series();
+            let rank_batch = Self::from_nonempty_columns(vec![rank_series])?;
+            return self.union(&rank_batch);
+        }
+
+        // Single row case - always rank 1
+        if self.len() == 1 {
+            rank_numbers[0] = 1;
+            let rank_series = UInt64Array::from((name.as_str(), rank_numbers)).into_series();
+            let rank_batch = Self::from_nonempty_columns(vec![rank_series])?;
+            return self.union(&rank_batch);
+        }
+
+        // Initialize rank tracking variables
+        let mut cur_rank = 1;
+        let mut next_rank = 1;
+
+        let order_by_table = self.eval_expression_list(order_by)?;
+
+        // Create a comparator for checking equality between rows
+        let comparator: Box<dyn Fn(usize, usize) -> bool + Send + Sync> =
+            build_multi_array_is_equal(
+                order_by_table.columns.as_slice(),
+                order_by_table.columns.as_slice(),
+                &vec![true; order_by_table.columns.len()],
+                &vec![true; order_by_table.columns.len()],
+            )?;
+
+        // First row always has rank 1
+        rank_numbers[0] = 1;
+
+        // Compute ranks for remaining rows
+        for (i, rank) in rank_numbers.iter_mut().enumerate().skip(1) {
+            // Always increment next_rank for regular rank()
+            if !dense {
+                next_rank += 1;
+            }
+
+            // Check if the current row has the same values as the previous row
+            let is_equal = comparator(i - 1, i);
+
+            if !is_equal {
+                // Different value, update rank
+                if dense {
+                    // For dense_rank, just increment by 1
+                    cur_rank += 1;
+                } else {
+                    // For rank(), use the next_rank which accounts for ties
+                    cur_rank = next_rank;
+                }
+            }
+
+            *rank = cur_rank;
+        }
+
+        let rank_series = UInt64Array::from((name.as_str(), rank_numbers)).into_series();
+        let rank_batch = Self::from_nonempty_columns(vec![rank_series])?;
+
+        self.union(&rank_batch)
+    }
+
+    // Helper method to process default values for window_offset
+    fn process_default_for_window(
```

Comment:
this function name is pretty unclear. also it's only used for offset right? maybe just define it in the window_offset function